### PR TITLE
Add Gradle tasks to configure and make host tools with CMake before compiling the Android Java bindings.

### PR DESCRIPTION
**Pull request by KoolJBlack**
Add Gradle tasks to configure and make host tools with CMake before compiling the Android Java bindings.

### CODE REVIEW

File: experimental/bindings/java/build.gradle [hunk 1]
```diff
@@ -54,8 +58,30 @@ android {
 
     externalNativeBuild {
         cmake {
-            version "3.13.0+"
```
Comment (ScottTodd) [inline experimental/bindings/java/build.gradle:57]: Why change to 3.18.4? We should match https://github.com/google/iree/blob/c67112d825feb8a8c820756f24ce6200028d2137/CMakeLists.txt#L15 unless there is a specific reason to require a higher version for gradle.

File: experimental/bindings/java/build.gradle [hunk 2]
```diff
@@ -54,8 +58,30 @@ android {
...
     }
 }
+
+// Task to cmake configure the host
+task cmakeConfigureHost(type: Exec) {
+    doFirst {
+        println "Configuring host tools with cmake..."
+    }
+    workingDir cmakeListDir
+    commandLine "cmake", "-G" , "Ninja",  "-B", hostBuildDir ,  "-DCMAKE_C_COMPILER=clang", "-DCMAKE_CXX_COMPILER=clang++",  "-DCMAKE_INSTALL_PREFIX=${hostInstallDir}", "."
```
Comment (ScottTodd) [inline experimental/bindings/java/build.gradle:73]: Can you drop the requirement to use clang? We support other compilers (MSVC in particular, but also gcc).

File: experimental/bindings/java/build.gradle [hunk 3]
```diff
@@ -54,8 +58,30 @@ android {
...
+            version "3.18.4+"
             path "../../../CMakeLists.txt"
         }
     }
 }
+
+// Task to cmake configure the host
+task cmakeConfigureHost(type: Exec) {
+    doFirst {
+        println "Configuring host tools with cmake..."
```
Comment (ScottTodd) [inline experimental/bindings/java/build.gradle:70]: If you're just building host tools, you could set more flags to prune the number of targets built.

https://github.com/google/iree/blob/c67112d825feb8a8c820756f24ce6200028d2137/build_tools/cmake/build_android.sh#L53-L59

In particular, set `-DIREE_BUILD_TESTS=OFF` so tests are not built or installed (those compile models and can take up to several minutes depending on the machine/config).

**Review by ScottTodd — COMMENTED** [review]
Cool :D less hacky than I was fearing (see https://github.com/google/iree/issues/4617, which this works around for the Java bindings)

File: experimental/bindings/java/build.gradle [hunk 4]
```diff
@@ -25,21 +29,21 @@ android {
...
-                        '-DIREE_BUILD_PYTHON_BINDINGS=OFF',
+                        "-DIREE_BUILD_COMPILER=OFF",
+                        "-DIREE_ENABLE_MLIR=OFF",
+                        "-DIREE_BUILD_TESTS=OFF",
+                        "-DIREE_BUILD_SAMPLES=OFF",
+                        "-DIREE_BUILD_PYTHON_BINDINGS=OFF",
 
                         // Run the commands at
                         // https://google.github.io/iree/get-started/getting-started-android-cmake#configure-and-build
                         // and update this flag to point to iree-build-host/install/.
```
Comment (not-jenni) [inline experimental/bindings/java/build.gradle:45]: This comment can be removed

File: experimental/bindings/java/build.gradle (hunk 1 above)
Comment (KoolJBlack) [reply experimental/bindings/java/build.gradle:57]: This was a mistake add. Reverted

File: experimental/bindings/java/build.gradle (hunk 2 above)
Comment (KoolJBlack) [reply experimental/bindings/java/build.gradle:73]: On my machine (linux) this doesn't compile unless you specify the compiler in this command. Is there another way to set it?

Comment (KoolJBlack) [reply experimental/bindings/java/build.gradle:73]: Scratch that. Looks like it works after exporting environment variables: https://stackoverflow.com/questions/45933732/how-to-specify-a-compiler-in-cmake/45934279#45934279

**Review by ScottTodd — APPROVED** [review]

**Review by not-jenni — APPROVED** [review]
Thanks Kojo!

File: experimental/bindings/java/build.gradle [hunk 5]
```diff
@@ -25,21 +29,17 @@ android {
 
         externalNativeBuild {
             cmake {
-                arguments '-DIREE_BUILD_EXPERIMENTAL_JAVA_BINDINGS=ON',
-                        '-DIREE_HAL_DRIVERS_TO_BUILD=VMLA',
+                arguments "-DIREE_BUILD_EXPERIMENTAL_JAVA_BINDINGS=ON",
+                        "-DIREE_HAL_DRIVERS_TO_BUILD=VMLA",
```
Comment (ScottTodd) [inline experimental/bindings/java/build.gradle:33]: We can also limit which components are built for the host tools with

```
        "-DIREE_TARGET_BACKENDS_TO_BUILD=vmla",
        "-DIREE_HAL_DRIVERS_TO_BUILD=vmla",
```

to match this. That will speed up builds significantly.

In full generality we should build all supported compiler target / drivers combinations (which would exclude Metal, CUDA, etc. on Android). For now the Java bindings only support VMLA (pending a way to create a device/driver of a specific type).

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**